finality-aleph/src/import.rs

```diff
@@ -77,82 +77,85 @@ where
 /// A wrapper around a block import that also extracts any present jsutifications and send them to
 /// our components which will process them further and possibly finalize the block.
 #[derive(Clone)]
-pub struct AlephBlockImport<B, I>
+pub struct AlephBlockImport<B, I, T>
 where
     B: BlockT,
     B::Header: Header<Number = BlockNumber>,
     I: BlockImport<B> + Clone + Send,
+    T: JustificationTranslator<B::Header>,
 {
     inner: I,
-    justification_tx: UnboundedSender<JustificationNotification<IdentifierFor<B>>>,
+    justification_tx: UnboundedSender<Justification<<B as BlockT>::Header>>,
+    translator: T,
 }
 
 #[derive(Debug)]
-enum SendJustificationError<B>
-where
-    B: BlockT,
-    B::Header: Header<Number = BlockNumber>,
-{
-    Send(TrySendError<JustificationNotification<IdentifierFor<B>>>),
+enum SendJustificationError<H: Header<Number = BlockNumber>, TE: Debug> {
+    Send(TrySendError<Justification<H>>),
     Consensus(Box<ConsensusError>),
     Decode(DecodeError),
+    Translate(TE),
 }
 
-impl<B> From<DecodeError> for SendJustificationError<B>
-where
-    B: BlockT,
-    B::Header: Header<Number = BlockNumber>,
+impl<H: Header<Number = BlockNumber>, TE: Debug> From<DecodeError>
+    for SendJustificationError<H, TE>
 {
     fn from(decode_error: DecodeError) -> Self {
         Self::Decode(decode_error)
     }
 }
 
-impl<B, I> AlephBlockImport<B, I>
+impl<B, I, T> AlephBlockImport<B, I, T>
 where
     B: BlockT,
     B::Header: Header<Number = BlockNumber>,
     I: BlockImport<B> + Clone + Send,
+    T: JustificationTranslator<B::Header>,
 {
     pub fn new(
         inner: I,
-        justification_tx: UnboundedSender<JustificationNotification<IdentifierFor<B>>>,
-    ) -> AlephBlockImport<B, I> {
+        justification_tx: UnboundedSender<Justification<B::Header>>,
+        translator: T,
+    ) -> AlephBlockImport<B, I, T> {
         AlephBlockImport {
             inner,
             justification_tx,
+            translator,
         }
     }
 
     fn send_justification(
         &mut self,
-        block_id: IdentifierFor<B>,
-        justification: Justification,
-    ) -> Result<(), SendJustificationError<B>> {
-        debug!(target: "aleph-justification", "Importing justification for block {:?}", block_id.number());
+        hash: B::Hash,
```

Comment:
Yes, but this conflicts with the translator abstraction. This will be fixed later, with the couple or PRs Maciek left us – we didn't want to merge them before, because they touch quite a lot of code that we are removing here.

There are currently actually a couple other ugly things we introduce in this PR that Maciek's refactors will/would have addressed. A bit of a bad timing, but likely the version with all this ugliness will never be run on any *net.